scripts/getSHAs.sh

```diff
@@ -62,14 +61,43 @@ getSHAs()
 		touch "$SHAs_FILE"
 	fi
 
+	# Get SHA for non-external tests by find .git dir
 	cd "$TEST_ROOT" || exit
 	find "$TEST_ROOT" -type d -name ".git" | while read -r gitDir; do
-		repoDir=$(dirname "$gitDir")
+		repoDir=$(realpath "$(dirname "$gitDir")")
 		cd "$repoDir" || continue
 		# append the info into $SHAs_FILR
 		{ echo "================================================"; echo "timestamp: $(timestamp)"; echo "repo dir: $repoDir"; echo "git repo: "; git remote show origin -n | grep "Fetch URL:"; echo "sha:"; git rev-parse HEAD; }  2>&1 | tee -a "$SHAs_FILE"
 	done
+	
+	cd "$TEST_ROOT" || exit
+	# Get SHA for external tests by test.properties
+	if [[ "$BUILD_LIST" == *"external"* ]]; then
+		for subDir in "$TEST_ROOT"/external/*/; do
+			 # find "$subDir" -type f -name 'Dockerfile.*'
+			if [[ $(find "$subDir" -type f -name 'Dockerfile.*') ]]; then
+				propertiesFile="$subDir/test.properties"
+				testDir=$(realpath "$subDir")			
+				if [[ -f "$propertiesFile" ]]; then
+					# read github_url and tag_version
+					github_url=$(grep '^github_url=' "$propertiesFile" | cut -d"=" -f2 | tr -d '"')
+					tag_version=$(grep '^tag_version=' "$propertiesFile" | cut -d"=" -f2 | tr -d '"')
+
+					if [[ -z "$github_url" || -z "$tag_version" ]]; then
+						echo "No github_url or tag_version in $propertiesFile"
+						continue
+					fi
+
+					# Retrieve the SHA using github_url and tag_version
+					sha=$(git ls-remote "$github_url" "refs/tags/$tag_version" | awk '{print $1}')
```

Comment:
Thanks @sophia-guo, is there ever a scenario where git ls-remote call will fail and if so, what is the fallback plan?

Reply:
Updated. Only do the tracking if sha is not empty, which scenario means the test will fail right at the beginning stage of setting up the environment and there is not any test output at all. Hence no need to track the sha.